### PR TITLE
added support for  huggingface dataset repositories

### LINKED ISSUE #1004: Kit import does not support huggingface dataset repositories
**Describe the problem you're trying to solve**
Kit import assumes huggingface repositories are regular models and attempts to download from the model URL (`https://huggingface.co/%s/resolve/%s/%s`). However, the URL format for dataset repositories is different, meaning `kit import` does not work for these repos at the moment.

**Additional context**
(none)

### PR DESCRIPTION
**Pull request by arnab2001**
### Description
add Hugging Face dataset support to kit import by detecting dataset URLs and hitting the correct api/datasets/resolve endpoints, so both models and datasets can be packaged

### Testing
`go test ./...`
`KITOPS_HOME=$(mktemp -d) ./kit import https://huggingface.co/datasets/nvidia/ProfBench --ref main --tag test/profbench:latest --tool hf`
`/kit unpack test/profbench:latest --dir "$(mktemp -d)"`

Screenshot 2025-11-04 at 6 04 19 PM

### Linked issues
closes #1004 

### AI-Assisted Code

- [x] This PR contains AI-generated code that I have reviewed and tested
- [x] I take full responsibility for all code in this PR, regardless of how it was created

### CODE REVIEW

File: pkg/cmd/kitimport/util.go [hunk 1]
```diff
@@ -163,20 +163,35 @@ func getEditorName() (string, error) {
...
+	var kind string
+	var normalizedPath string
+
+	if len(segments) == 3 && segments[0] == "datasets" {
+		kind = "dataset"
+		normalizedPath = strings.Join(segments[1:], "/")
+	} else if len(segments) == 2 {
+		if segments[0] == "datasets" {
+			return "", "", fmt.Errorf("invalid dataset URL format: expected datasets/org/repo, got '%s'", path)
+		}
```
Comment (amisevsk) [inline pkg/cmd/kitimport/util.go]: I don't know if this check is appropriate for _all_ uses of `extractRepoFromURL` -- it makes sense for huggingface URLs but `kit import` supports any git repository as well; what if I try to import from `my-gitlab.org/datasets/my-dataset`? 

This is mixing huggingface-specific logic into this function without making it clear; this function should handle both git repositories (arbitrary names) and huggingface ones. 

We should either handle huggingface specifically or rename the function to reflect its actual purpose and update all usages to be appropriate.

File: pkg/cmd/kitimport/util.go [hunk 2]
```diff
@@ -163,20 +163,35 @@ func getEditorName() (string, error) {
...
-		return "", fmt.Errorf("failed to parse url: %w", err)
+		return "", "", fmt.Errorf("failed to parse url: %w", err)
 	}
 
 	path := strings.Trim(u.Path, "/")
 	segments := strings.Split(path, "/")
-	if len(segments) != 2 {
-		return "", fmt.Errorf("could not extract organization and repository from '%s'", path)
+
+	var kind string
```
Comment (amisevsk) [inline pkg/cmd/kitimport/util.go]: We should probably use a typed enum for `kind` to make things more explicit

```go
type hfRepoType int

const (
    hfModelRepoType hfRepoType = iota
    hfDatasetRepoType
)
```

File: pkg/cmd/kitimport/util.go [hunk 3]
```diff
@@ -163,20 +163,35 @@ func getEditorName() (string, error) {
...
-// Returns an error we cannot automatically handle the input URL/string.
-//   - https://example.com/segment1/segment2 --> segment1/segment2
-//   - 'organization/repository'             --> 'organization/repository'
-func extractRepoFromURL(rawUrl string) (string, error) {
+// Returns the normalized repository path (org/repo), the kind ("model" or "dataset"), and an error if parsing fails.
+//   - https://huggingface.co/org/repo --> (org/repo, "model", nil)
+//   - https://huggingface.co/datasets/org/repo --> (org/repo, "dataset", nil)
+//   - 'organization/repository' --> (organization/repository, "model", nil)
+//   - 'datasets/organization/repository' --> (organization/repository, "dataset", nil)
+func extractRepoFromURL(rawUrl string) (string, string, error) {
```
Comment (amisevsk) [inline pkg/cmd/kitimport/util.go]: These changes change the purpose of this function, making it specific for handling huggingface repositories.

```go
// extractRepoFromURL attempts to normalize a string or URL into a repository name as is used on GitHub and Huggingface.
```
vs
```
Returns the normalized repository path (org/repo), the kind ("model" or "dataset"),
```
-- the 'kind' here only makes sense in a huggingface context, not GitHub (or any Git repository, for that matter).

**Comment by arnab2001** [comment]
Thanks for the review @amisevsk, Its my short sight that i missed these things.  I am working on resolving these comments

File: pkg/cmd/kitimport/util.go (hunk 2 above)
Comment (arnab2001) [reply pkg/cmd/kitimport/util.go]: Both enums use int with iota as suggested
mapRepoType() converts between the two enum types

File: pkg/cmd/kitimport/util.go [hunk 4]
```diff
@@ -163,20 +163,35 @@ func getEditorName() (string, error) {
...
-	if len(segments) != 2 {
-		return "", fmt.Errorf("could not extract organization and repository from '%s'", path)
+
+	var kind string
+	var normalizedPath string
+
+	if len(segments) == 3 && segments[0] == "datasets" {
+		kind = "dataset"
+		normalizedPath = strings.Join(segments[1:], "/")
+	} else if len(segments) == 2 {
```
[the remark this answers is not in the document]
Comment (arnab2001) [reply pkg/cmd/kitimport/util.go]: created Typed enums called RepositoryType and hfRepoType ,No string literals like "dataset" or "model" used in comparisons

**Comment by arnab2001** [comment]
- Created typed enums: RepositoryType in pkg/lib/hf/repo.go and hfRepoType in pkg/cmd/kitimport/hfimport.go, now no string literals ("dataset", "model")  are used in comparisons.
All comparisons use typed constants now

- Seperation: extractRepoFromURL() in util.go is now generic and handles GitHub/Git URLs only, parseHuggingFaceRepo() in hfimport.go handles all HuggingFace-specific logic (including dataset detection). extractRepoFromURL() is no longer mixing concerns

- Both enums use int with iota as suggested
Proper mapping function mapRepoType() converts between the two enum types

@amisevsk  please check and let me know if theres any other changes are required, thank you for your time.

File: pkg/cmd/kitimport/util.go (hunk 3 above)
Comment (arnab2001) [reply pkg/cmd/kitimport/util.go]: extractRepoFromURL() remains generic and only returns the normalized repo path (org/repo) , no kind/type. HuggingFace-specific logic (including dataset detection) is handled by parseHuggingFaceRepo() in hfimport.go.

File: pkg/cmd/kitimport/util.go (hunk 2 above)
Comment (arnab2001) [reply pkg/cmd/kitimport/util.go]: Implemented. Added hfRepoType enum in hfimport.go (lines 178-183) and RepositoryType in pkg/lib/hf/repo.go. All comparisons use typed constants (RepoTypeDataset, RepoTypeModel) instead of string literals.

File: pkg/cmd/kitimport/util.go (hunk 1 above)
Comment (arnab2001) [reply pkg/cmd/kitimport/util.go]: extractRepoFromURL() is generic now and works for any git repository (GitLab, GitHub, etc.). my-gitlab.org/datasets/my-dataset would be handled correctly by extractRepoFromURL() as a generic org/repo pair. HuggingFace-specific parsing (including "datasets" prefix handling) is isolated in parseHuggingFaceRepo(). Usage in cmd.go (lines 134-142) tries HuggingFace parsing first, then falls back to generic parsing.

**Comment by gorkem** [comment]
PR checks are failing

**Comment by arnab2001** [comment]
@gorkem  working on it

**Comment by arnab2001** [comment]
@gorkem , CI checks should pass now

File: pkg/cmd/kitimport/hfimport.go [hunk 5]
```diff
@@ -172,3 +174,46 @@ func kitfileHasCatchallLayer(kitfile *artifact.KitFile) bool {
...
+	if len(segments) == 2 && segments[0] == "datasets" {
+		return "", 0, fmt.Errorf("could not extract repository from path '%s'", path)
+	}
+
+	if len(segments) >= 2 {
+		return strings.Join(segments[len(segments)-2:], "/"), hfRepoTypeModel, nil
+	}
+
+	return "", 0, fmt.Errorf("could not extract repository from path '%s'", path)
+}
```
Comment (amisevsk) [inline pkg/cmd/kitimport/hfimport.go]: We should move this function into `lib/hf` and have it return the `lib/hf`-defined repo types (e.g. it could go into `repo.go`. This would allow us to avoid the duplicate `hfRepoType` enum and the `mapRepoType` function.

File: pkg/cmd/kitimport/hfimport.go [hunk 6]
```diff
@@ -172,3 +174,46 @@ func kitfileHasCatchallLayer(kitfile *artifact.KitFile) bool {
...
+		return hf.RepoTypeDataset
+	default:
+		return hf.RepoTypeModel
+	}
+}
+
+func parseHuggingFaceRepo(rawURL string) (string, hfRepoType, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return "", 0, fmt.Errorf("failed to parse url: %w", err)
```
Comment (amisevsk) [inline pkg/cmd/kitimport/hfimport.go]: Consider defining an explicit `repoTypeUnknown` in the enum (placing it first in the list) and handling that in the `lib/hf` functions to improve zero-values for the error case.

File: pkg/cmd/kitimport/hfimport.go [hunk 7]
```diff
@@ -172,3 +174,46 @@ func kitfileHasCatchallLayer(kitfile *artifact.KitFile) bool {
...
+	}
+
+	if u.Host != "" && !strings.Contains(u.Host, "huggingface.co") {
+		return "", 0, fmt.Errorf("not a Hugging Face repository")
+	}
+
+	path := strings.Trim(u.Path, "/")
+	segments := strings.FieldsFunc(path, func(r rune) bool { return r == '/' })
```
Comment (amisevsk) [inline pkg/cmd/kitimport/hfimport.go]: Since you're trimming `/` from the path, FieldsFunc will be equivalent to `strings.Split(path, "/")` except harder to read.

**Comment by arnab2001** [comment]
@amisevsk  Done. Moved ParseHuggingFaceRepo to pkg/lib/hf/repo.go, added RepoTypeUnknown as the zero value, and simplified FieldsFunc to strings.Split. Removed the duplicate hfRepoType enum and mapRepoType function from hfimport.go.

File: pkg/lib/hf/repo.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+// ParseHuggingFaceRepo parses a Hugging Face repository URL or path and returns
+// the normalized repository path (org/repo) and the repository type.
+func ParseHuggingFaceRepo(rawURL string) (string, RepositoryType, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return "", RepoTypeUnknown, fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	if u.Host != "" && !strings.Contains(u.Host, "huggingface.co") {
```
Comment (gorkem) [inline pkg/lib/hf/repo.go]: In this check, inputs like https://huggingface.co.evil.com/org/repo pass the “is HF” check and proceed as if they were Hugging Face. Can you tighten the parsing and validation of the host?

File: pkg/lib/hf/repo.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		return "", RepoTypeUnknown, fmt.Errorf("not a Hugging Face repository")
+	}
+
+	path := strings.Trim(u.Path, "/")
+	segments := strings.Split(path, "/")
+
+	if len(segments) >= 3 && segments[0] == "datasets" {
```
Comment (gorkem) [inline pkg/lib/hf/repo.go:58]: `kit import huggingface.co/datasets/org/repo --tool hf` will fail  because `huggingface.co/datasets/org/repo` **(no scheme)** will be treated as RepoTypeModel because when the scheme is missing segments[0] is the Host (aka huggingface.co)

**Comment by arnab2001** [comment]
Working on it

**Comment by arnab2001** [comment]
@gorkem 

Fixed two critical issues in ParseHuggingFaceRepo(): hardened host validation to reject malicious URLs like https://huggingface.co.evil.com using strict exact-match checks, and added scheme-less URL handling so huggingface.co/datasets/org/repo correctly parses as a dataset instead of failing.

Added 3 new test cases covering security edge cases, trailing slashes, and HTTP schemes - all 12 tests passing.

**Review by amisevsk — APPROVED** [review]
I realized today I had left this PR on my TODO list for too long; sorry for the delay @arnab2001 and thank you for your contribution :+1:
